File: typescript/packages/core/src/commands/cli/builtin/git/pickaxe.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import git from 'isomorphic-git'

import { repoArgs, type Repo } from './repo.ts'
import { treeEntries } from './tree.ts'

const DEC = new TextDecoder('utf-8', { fatal: false })

/** How many times a string appears in one blob. */
async function occurrences(repo: Repo, oid: string | null, needle: string): Promise<number> {
  if (oid === null) return 0
  let data: Uint8Array
  try {
    data = (await git.readBlob({ ...repoArgs(repo), oid })).blob
  } catch {
    return 0
  }
  const text = DEC.decode(data)
  if (needle === '') return 0
  let count = 0
  let at = text.indexOf(needle)
  while (at !== -1) {
    count += 1
    at = text.indexOf(needle, at + needle.length)
  }
  return count
}

/**
 * Whether a commit changed the number of occurrences of a string.
 *
 * This is git's `-S` (pickaxe), and it is deliberately not a grep: a commit that
 * merely moves a line containing the string does not change how many times the
 * string appears, so it is not reported. The commit that *introduced* the string
 * is, which is what makes `-S <name> --reverse` answer "where did this come
 * from".
 *
 * Compared against the first parent, or against nothing for a root commit, so
 * the objects a root commit adds all count as introduced.
 */
export async function touches(
  repo: Repo,
  oid: string,
  parents: readonly string[],
  needle: string,
): Promise<boolean> {
  const { commit } = await git.readCommit({ ...repoArgs(repo), oid })
  const after = await treeEntries(repo, commit.tree)
  const first = parents[0]
  let before = new Map<string, { oid: string; mode: string }>()
  if (first !== undefined) {
    const parent = await git.readCommit({ ...repoArgs(repo), oid: first })
    before = await treeEntries(repo, parent.commit.tree)
  }
  for (const path of new Set([...before.keys(), ...after.keys()])) {
    const old = before.get(path)?.oid ?? null
    const now = after.get(path)?.oid ?? null
    if (old === now) continue
    if ((await occurrences(repo, old, needle)) !== (await occurrences(repo, now, needle))) {
      return true
    }
  }
  return false
}

File: python/mirage/commands/cli/builtin/git/pickaxe.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from dulwich.diff_tree import tree_changes
from dulwich.object_store import BaseObjectStore
from dulwich.objects import Blob, Commit, ObjectID

EMPTY_TREE = None


def _occurrences(store: BaseObjectStore, sha: ObjectID | None,
                 needle: bytes) -> int:
    """How many times a string appears in one blob.

    Args:
        store (BaseObjectStore): object database holding the blob.
        sha (ObjectID | None): blob id, None when the side does not
            exist.
        needle (bytes): the string being counted.
    """
    if sha is None:
        return 0
    obj = store[sha]
    if not isinstance(obj, Blob):
        return 0
    return obj.data.count(needle)


def touches(store: BaseObjectStore, commit: Commit, needle: bytes) -> bool:
    """Whether a commit changed the number of occurrences of a string.

    This is git's ``-S`` (pickaxe), and it is deliberately not a grep: a
    commit that merely moves a line containing the string does not
    change how many times the string appears, so it is not reported. The
    commit that *introduced* the string is, which is what makes
    ``-S <name> --reverse`` answer "where did this come from".

    Compared against the first parent, or against nothing for a root
    commit, so the objects a root commit adds all count as introduced.

    Args:
        store (BaseObjectStore): object database holding the trees.
        commit (Commit): the commit to test.
        needle (bytes): the string being counted.
    """
    parent_tree = EMPTY_TREE
    if commit.parents:
        parent = store[commit.parents[0]]
        assert isinstance(parent, Commit)
        parent_tree = parent.tree
    for change in tree_changes(store, parent_tree, commit.tree):
        old = change.old.sha if change.old is not None else None
        new = change.new.sha if change.new is not None else None
        if _occurrences(store, old,
                        needle) != _occurrences(store, new, needle):
            return True
    return False
